### PR TITLE
Configuration of trust stores for Prometheus and Apicurio clients

### LINKED ISSUE #1277: Support trusted certificate configuration for metrics sources and schema registries
Allow for a trusted certificate to be configured for metrics endpoints and schema registries

### PR DESCRIPTION
**Pull request by MikeEdgar**
Closes #1277 

--- 

Approach

This change allows for a trust store to be configured for metrics source (Prometheus) and schema registry (Apicurio) clients. The trust store may be set in the Console CR along with the URLs for each connection type.

For example:
```yaml
...
spec:
  metricsSources:
    - name: ocp-prometheus
      type: openshift-monitoring
      url: https://thanos-querier.mycluster.example.com
      truststore:
        type: PEM  # also supports JKS or PKCS12
        alias: "" # optional, only for JKS or PKCS12
        password: {} # optional, only for JKS or PKCS12. Same structure as `content` below
        content:
          # only one of `value` or `valueFrom` is used
          value: |
            -----BEGIN CERTIFICATE-----
            MIIFazCCA1OgAwIBAgIRAIIQz7DSQONZRGPgu2OCiwAwDQYJKoZIhvcNAQELBQAw
            ...
            emyPxgcYxn/eR44/KJ4EBs+lVDR3veyJm+kXQ99b21/+jh5Xos1AnX5iItreGCc=
            -----END CERTIFICATE-----
          # May use a ConfigMap or Secret key when no literal value is given
          valueFrom:
            configMapKeyRef: {}
            secretKeyRef: {}
...
```

The operator will then read the trust store content either directly from the CR for literal values or from the ConfigMap or Secret, copy to the Console secret being reconciled, and create a volume/mount and appropriate Quarkus environment vars for a TLS configuration bucket [1] specific to the connection. 

The clients in the API server will then obtain the trust store/SSL context from the TLS bucket named by convention and use for the connection to the remote service.

[1] https://quarkus.io/guides/tls-registry-reference#using-the-tls-registry

### CODE REVIEW

**Comment by MikeEdgar** [comment]
@katheris , @k-wall this PR still needs some unit/integration tests, but I am curious to get your thoughts on the general approach/pattern to implement this.

File: api/src/main/java/com/github/streamshub/console/api/service/MetricsService.java
Comment (MikeEdgar) [inline api/src/main/java/com/github/streamshub/console/api/service/MetricsService.java:123]: The name may be slightly different (dots versus dashes) depending on whether the property was set via a system property or an environment variable.

File: operator/src/main/java/com/github/streamshub/console/dependents/ConsoleSecret.java
Comment (MikeEdgar) [inline operator/src/main/java/com/github/streamshub/console/dependents/ConsoleSecret.java:202]: Ultimately, I'd like to avoid configuration of these buckets this way and have it supported via the console's YAML file, but that will be a bit too involved for the initial support I think.

**Comment by k-wall** [comment]
>         alias: "" # optional, only for JKS or PKCS12

what's the idea here?  I'm familiar with alias on the keystore side -  to select a single key from potentially many in the store, but in the truststore side you normally trust all the certificates in the truststore.  In Java TrustManagerFactory accepts a truststore and gives you a TrustManager which trusts them all.  So, what's the role of the alias?

**Comment by k-wall** [comment]
>         password: {} # optional, only for JKS or PKCS12. Same structure as `content` below

is this a ref to a secret?  you would normally try to avoid an inline secret.

**Comment by MikeEdgar** [comment]
> > ```
> >     alias: "" # optional, only for JKS or PKCS12
> > ```
> 
> what's the idea here? I'm familiar with alias on the keystore side - to select a single key from potentially many in the store, but in the truststore side you normally trust all the certificates in the truststore. In Java TrustManagerFactory accepts a truststore and gives you a TrustManager which trusts them all. So, what's the role of the alias?

That's what I've typically seen also, but the TLS registry supports configuration of an `alias` for JKS and P12 [1]. I figured better to align to the target on it.

[1] https://quarkus.io/guides/tls-registry-reference#trust-stores

**Comment by MikeEdgar** [comment]
> > ```
> >     password: {} # optional, only for JKS or PKCS12. Same structure as `content` below
> > ```
> 
> is this a ref to a secret? you would normally try to avoid an inline secret.

Yes, it can be. The field allows for `value` or `valueFrom` like the truststore `content`. I kept the structure the same to allow security, but also convenience for development and testing, etc.

File: api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java
Comment (k-wall) [inline api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java:101]: Oh my!

File: api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java
Comment (k-wall) [inline api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java:37]: Do we have a path to removing this code from the Console?

File: operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java
Comment (k-wall) [inline operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java:67]: I don't understand why `coalesce` is called coalesce.  it seems to be `valueOrDefault()`

File: operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java
Comment (k-wall) [inline operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java:101]: the approach looks worrying.   doesn't this potentially overwrite env vars coming from the `console.deployment.yaml` with values provided by the user from the CR.

File: operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java
Comment (MikeEdgar) [reply operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java:67]: This is my SQL history coming in :laughing: ... it's also named something similar in JavaScript [1]. Functionally it's `valueOrDefault`, but less verbose.

[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing

File: api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java
Comment (MikeEdgar) [reply api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java:37]: I haven't given too much thought yet, other than to think that this kind of support should be in the Quarkus extension [1] somehow.

[1] https://github.com/quarkusio/quarkus/tree/main/extensions/schema-registry/apicurio

File: operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java
Comment (MikeEdgar) [reply operator/src/main/java/com/github/streamshub/console/dependents/ConsoleDeployment.java:101]: That's a good point, the ENVs in the operator should always be present. If user attempts to override it results in a deployment error because they are duplicates. That is probably the correct behavior, but it happened to me trying to debug this and I added this remove line in haste. I'll go ahead and take it out.

File: api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java
Comment (k-wall) [reply api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java:37]: Yes, I think it is worth separately opening a RFE/putting up a PR to get this code back where it belongs.  Although maybe it is worth holding off until the dust from Apicurio 3.0 has settled.

**Comment by k-wall** [comment]
On the whole, the approach looks good.  I'm concerned about some of the coding in the ApicurioClient but will be content to overlook providing there's a plan to address the debt.

File: api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java
Comment (MikeEdgar) [reply api/src/main/java/com/github/streamshub/console/api/support/serdes/ApicurioClient.java:37]: I've opened https://github.com/quarkusio/quarkus/issues/45069 to track this.

**Comment by MikeEdgar** [comment]
@k-wall I'll go ahead and merge this if no objections. The config option in the CR will be used also for #1248 to support a TLS truststore for the OIDC provider (building on #1126 once that's merged)

**Review by k-wall — APPROVED** [review]

**Pull request merged**
